File: source/graphics/mesh.cpp
#include "../../include/graphics/mesh.h"
#include "../../include/core/console.h"

#include <glad/glad.h>
#include <iterator>

Mesh::Mesh() {
    vao = 0;
    vbo = 0;
    ebo = 0;
    count = 0;
    primitiveType = GL_TRIANGLES;
    drawType = GL_STATIC_DRAW;
}
Mesh::Mesh(
    std::vector<float> vertices,
    std::vector<VertexAttribute> attributes,
    GLenum primitiveType,
    GLenum drawType) {

    this->primitiveType = primitiveType;
    this->drawType = drawType;
    set_mesh(vertices, attributes);
}
Mesh::Mesh(
    std::vector<float> vertices,
    std::vector<unsigned int> indices,
    std::vector<VertexAttribute> attributes,
    GLenum primitiveType,
    GLenum drawType) {

    this->primitiveType = primitiveType;
    this->drawType = drawType;
    set_mesh(vertices, indices, attributes);
}
Mesh* Mesh::set_mesh(std::vector<float> vertices, std::vector<VertexAttribute> attributes) {
    if (vertices.size() <= 0)
        return nullptr;

    usingIndices = false;

    //generate vao, vbo, and ebo
    if (!generatedBuffers) {
        glGenVertexArrays(1, &vao);
        glGenBuffers(1, &vbo);
    }

    //bind vao
    glBindVertexArray(vao);

    //assign vbo data
    count = vertices.size();

    glBindBuffer(GL_ARRAY_BUFFER, vbo);
    glBufferData(GL_ARRAY_BUFFER, count * sizeof(float), &vertices[0], drawType);

    //create vertex attributes
    unsigned int stride = VertexAttribute::get_total_size_in_bytes(attributes);
    unsigned int offset = 0;
    for (int i = 0; i < attributes.size(); ++i)
    {
        VertexAttribute attrib = attributes[i];

        glVertexAttribPointer(i, attrib.componentsCount, attrib.pointerType, GL_FALSE, stride, (void*)offset);
        glEnableVertexAttribArray(i);
        offset += attrib.sizeInBytes;
    }

    generatedBuffers = true;
    return this;
}
Mesh* Mesh::set_mesh(std::vector<float> vertices, std::vector<unsigned int> indices, std::vector<VertexAttribute> attributes) {
    if (vertices.size() <= 0 || indices.size() <= 0)
        return nullptr;

    usingIndices = true;

    //generate vao, vbo, and ebo
    if (!generatedBuffers) {
        glGenVertexArrays(1, &vao);
        glGenBuffers(1, &vbo);
        glGenBuffers(1, &ebo);
    }

    //bind vao
    glBindVertexArray(vao);

    //assign vbo data
    glBindBuffer(GL_ARRAY_BUFFER, vbo);
    glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(float), &vertices[0], drawType);

    //assign ebo data
    count = indices.size();

    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, ebo);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, count * sizeof(unsigned int), &indices[0], drawType);

    //create vertex attributes
    unsigned int stride = VertexAttribute::get_total_size_in_bytes(attributes);
    unsigned int offset = 0;
    for (int i = 0; i < attributes.size(); ++i)
    {
        VertexAttribute attrib = attributes[i];

        glVertexAttribPointer(i, attrib.componentsCount, attrib.pointerType, GL_FALSE, stride, (void*)offset);
        glEnableVertexAttribArray(i);
        offset += attrib.sizeInBytes;
    }

    generatedBuffers = true;
    return this;
}
void Mesh::draw() {
    if (!generatedBuffers)
        return;

    glBindVertexArray(vao);
    if (usingIndices)
        glDrawElements(primitiveType, count, GL_UNSIGNED_INT, 0);
    else
        glDrawArrays(primitiveType, 0, count);
    glBindVertexArray(0);
}
void Mesh::destroy() {
    glDeleteBuffers(1, &vao);
    glDeleteBuffers(1, &vbo);
    glDeleteBuffers(1, &ebo);
}

File: CSharp/ASE_CSharp/Graphics/Mesh.cs
using OpenTK.Graphics.OpenGL4;

namespace ASE.Graphics
{
    public class Mesh : IDisposable
    {
        public uint vao;
        public uint vbo;
        public uint ebo;
        public int count;
        public PrimitiveType primitiveType;
        public BufferUsageHint drawType;

        private bool generatedBuffers;
        private bool usingIndices;

        public Mesh()
        {
            
        }
        public Mesh(float[] vertices, PrimitiveType primitiveType, BufferUsageHint drawType, params VertexAttribute[] attributes)
        {
            this.primitiveType = primitiveType;
            this.drawType = drawType;
            SetMesh(vertices, attributes);
        }
        public Mesh(float[] vertices, int[] indices, PrimitiveType primitiveType, BufferUsageHint drawType, params VertexAttribute[] attributes)
        {
            this.primitiveType = primitiveType;
            this.drawType = drawType;
            SetMesh(vertices, indices, attributes);
        }
        public Mesh(List<float> vertices, PrimitiveType primitiveType, BufferUsageHint drawType, params VertexAttribute[] attributes)
        {
            this.primitiveType = primitiveType;
            this.drawType = drawType;
            SetMesh(vertices.ToArray(), attributes);
        }
        public Mesh(List<float> vertices, List<int> indices, PrimitiveType primitiveType, BufferUsageHint drawType, params VertexAttribute[] attributes)
        {
            this.primitiveType = primitiveType;
            this.drawType = drawType;
            SetMesh(vertices.ToArray(), indices.ToArray(), attributes);
        }
        public Mesh SetMesh(float[] vertices, params VertexAttribute[] attributes)
        {
            if (vertices.Length <= 0)
            {
                Console.WriteLine("Attempted to create mesh with no data. This is not allowed.");
                return this;
            }

            usingIndices = false;

            if (!generatedBuffers)
            {
                GL.GenVertexArrays(1, out vao);
                GL.GenBuffers(1, out vbo);
            }

            count = vertices.Length;

            GL.BindVertexArray(vao);

            GL.BindBuffer(BufferTarget.ArrayBuffer, vbo);
            GL.BufferData(BufferTarget.ArrayBuffer, count * sizeof(float), vertices.ToArray(), drawType);

            int stride = VertexAttribute.GetTotalSizeInBytes(attributes);
            int offset = 0;
            for (int i = 0; i < attributes.Length; i++)
            {
                VertexAttribute attrib = attributes[i];

                GL.VertexAttribPointer(i, attrib.componentsCount, attrib.pointerType, false, stride, offset);
                GL.EnableVertexAttribArray(i);
            }

            generatedBuffers = true;
            return this;
        }
        public Mesh SetMesh(float[] vertices, int[] indices, params VertexAttribute[] attributes)
        {
            if (vertices.Length <= 0 || indices.Length <= 0)
            {
                Console.WriteLine("Attempted to create mesh with no data. This is not allowed.");
                return this;
            }

            usingIndices = true;

            if (!generatedBuffers)
            {
                GL.GenVertexArrays(1, out vao);
                GL.GenBuffers(1, out vbo);
                GL.GenBuffers(1, out ebo);
            }

            GL.BindVertexArray(vao);

            GL.BindBuffer(BufferTarget.ArrayBuffer, vbo);
            GL.BufferData(BufferTarget.ArrayBuffer, vertices.Length * sizeof(float), vertices.ToArray(), drawType);

            count = indices.Length;

            GL.BindBuffer(BufferTarget.ElementArrayBuffer, ebo);
            GL.BufferData(BufferTarget.ElementArrayBuffer, count * sizeof(uint), indices.ToArray(), drawType);

            int stride = VertexAttribute.GetTotalSizeInBytes(attributes);
            int offset = 0;
            for (int i = 0; i < attributes.Length; i++)
            {
                VertexAttribute attrib = attributes[i];

                GL.VertexAttribPointer(i, attrib.componentsCount, attrib.pointerType, false, stride, offset);
                GL.EnableVertexAttribArray(i);
            }

            generatedBuffers = true;
            return this;
        }
        //draw normally
        public void Draw()
        {
            if (!generatedBuffers || count == 0)
                return;

            GL.BindVertexArray(vao);
            if (usingIndices)
                GL.DrawElements(primitiveType, count, DrawElementsType.UnsignedInt, 0);
            else
                GL.DrawArrays(primitiveType, 0, count);

            GL.BindVertexArray(0);
        }
        //draw override with special primitive type
        public void Draw(PrimitiveType primitiveType)
        {
            if (!generatedBuffers)
                return;

            GL.BindVertexArray(vao);
            if (usingIndices)
                GL.DrawElements(primitiveType, count, DrawElementsType.UnsignedInt, 0);
            else
                GL.DrawArrays(primitiveType, 0, count);

            GL.BindVertexArray(0);
        }
        public void Clear()
        {
            count = 0;
        }
        private bool _disposed;
        ~Mesh()
        {
            Dispose(false);
        }
        public void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                GL.DeleteBuffers(1, ref vao);
                GL.DeleteBuffers(1, ref vbo);
                GL.DeleteBuffers(1, ref ebo);
                _disposed = true;
            }
        }
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
    }
}
